classes/merge_request.php

```diff
@@ -0,0 +1,170 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Version information
+ *
+ * @package     tool
+ * @subpackage  mergeusers
+ * @author      Nicola Vallinoto, Liguria Digitale
+ * @author      Jordi Pujol-Ahulló, SREd, Universitat Rovira i Virgili
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace tool_mergeusers;
+class merge_request {
+    /**
+     * Missing merge request id. This id really does not exist.
+     * @var integer
+     */
+    const MISSING_MERGE_REQUEST_ID = 1;
+    /**
+     * Merge request queued but not yet processed.
+     * @var integer
+     */
+    const QUEUED_NOT_PROCESSED = 2;
+    /**
+     * Merge request queued, to be processed (the adhoc task is created for it).
+     * @var integer
+     */
+    const QUEUED_TO_BE_PROCESSED = 3;
+    /**
+     * Merge request is in progress but not concluded (its adhoc_task has started).
+     * @var integer
+     */
+    const INPROGRESS_NOT_CONCLUDED = 4;
+    /**
+     * Tried with error; pending of retrial (the merge process has found some error
+     * and the adhoc_task is queued to be retried in the future).
+     * @var integer
+     */
+    const TRIED_WITH_ERROR = 5;
+    /**
+     * Completed with success.
+     * @var integer
+     */
+    const COMPLETED_WITH_SUCCESS = 6;
+    /**
+     * Completed with errors (even with all the current retries the merge process
+     * could not end with success).
+     * @var integer
+     */
+    const COMPLETED_WITH_ERRORS = 7;
+    /**
+     * Reference table for merge requests.
+     * @var string
+     */
+    const TABLE_MERGE_REQUEST = 'tool_mergeusers_queue';
+     /**
+     * Reference table for merge requests. Before web service integration.
+     * @var string
+     */
+    const TABLE_MERGE_REQUEST_OLD = 'tool_mergeusers';
+    /**
+     * Reference table for users.
+     * @var string
+     */
+    const TABLE_USERS = 'user';
+    private $data;
+    private function __construct(stdClass $data) {
+        $this->data = $data;
+    }
+
+    public static function from(stdClass $data) {
+        return new self($data);
+    }
+
+    public function __get($name) {
+        if (isset($this->data->${name})) {
+            $value = $this->data->${name};
+            if ($name == 'log') {
+                $value = json_decode($value, true);
+            }
+            return $value;
+        }
+        return null;
+    }
+    public function __set($name, $value) {
+        if (isset($this->data->${name})) {
+            if ($name == 'log') {
+                $value = json_encode($value);
+            }
+            $this->data->${name} = $value;
+        }
+    }
+
```

Comment:
I would add a method like:

```php
public function append_log(array $newlog, int $logtime) {
     $log = json_decode($this->data->log, true);
     $log[$logtime] = $newlog;
     $this->data->log = json_encode($log);
}
```

This would help us to deal with appending logs wherever it is necessary in this patch.

Reply:
Why you use $this->data->log and not a return type as json_encode($log) ?

Reply:
I am not following you, sorry. This methos abstracts how to array log is organized internally in the merge_request class.

Here, I assume that the field $this->data->log is always in json format. This is way firstly, we have to decode the potentially existing logs into the array, append the new log, and then store the whole new log array in json format.

It is the whole array that is converted into json format, not only each log entry.

Reply:
And in this method, we are not returning anything. We are just appending new log entries into the merge_request logs.

Reply:
I updated the signature of the method to make it explicit, to inform that nothing is returned.